File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;

class Solution {
    private Set<String> resultSet = new HashSet<>();
    
    public int solution(String[] user_id, String[] banned_id) {
        boolean[] visited = new boolean[user_id.length];
        dfs(user_id, banned_id, visited, 0);
        return resultSet.size();
    }
    
    private void dfs(String[] user_id, String[] banned_id, boolean[] visited, int depth) {
        if (depth == banned_id.length) {
            StringBuilder combination = new StringBuilder();
            for (int i = 0; i < user_id.length; i++) {
                if (visited[i]) {
                    combination.append(i);
                }
            }
            resultSet.add(combination.toString());
            return;
        }

        for (int i = 0; i < user_id.length; i++) {
            if (!visited[i] && isMatch(user_id[i], banned_id[depth])) {
                visited[i] = true;
                dfs(user_id, banned_id, visited, depth + 1);
                visited[i] = false;
            }
        }
    }
    
    private boolean isMatch(String userId, String bannedId) {
        if (userId.length() != bannedId.length()) {
            return false;
        }
        for (int i = 0; i < userId.length(); i++) {
            if (bannedId.charAt(i) != '*' && bannedId.charAt(i) != userId.charAt(i)) {
                return false;
            }
        }
        
        return true;
    }

}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
def solution(user_id, banned_id):
    result_set = set()

    def is_match(user, banned):
        if len(user) != len(banned):
            return False
        for u, b in zip(user, banned):
            if b != '*' and b != u:
                return False
        return True

    def dfs(visited, depth):
        if depth == len(banned_id):
            result_set.add(tuple(visited))
            return

        for i in range(len(user_id)):
            if not visited[i] and is_match(user_id[i], banned_id[depth]):
                visited[i] = True
                dfs(visited, depth + 1)
                visited[i] = False

    dfs([False] * len(user_id), 0)
    
    return len(result_set)